modules/spotxBidAdapter.js

```diff
@@ -0,0 +1,330 @@
+import * as utils from 'src/utils';
+import { Renderer } from 'src/Renderer';
+import { registerBidder } from 'src/adapters/bidderFactory';
+import { BANNER, VIDEO } from 'src/mediaTypes';
+
+const BIDDER_CODE = 'spotx';
+const URL = '//search.spotxchange.com/openrtb/2.3/dados/';
+const ORTB_VERSION = '2.3';
+
+export const spec = {
+  code: BIDDER_CODE,
+  aliases: ['spotx'],
+  supportedMediaTypes: [BANNER, VIDEO],
+
+  /**
+   * Determines whether or not the given bid request is valid.
+   * From Prebid.js: isBidRequestValid - Verify the the AdUnits.bids, respond with true (valid) or false (invalid).
+   *
+   * @param {object} bid The bid to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+   */
+  isBidRequestValid: function(bid) {
+    if (bid && (typeof bid.params !== 'object' || typeof bid.params.video !== 'object')) {
+      utils.logMessage(BIDDER_CODE + ': video params is missing or is incorrect');
+      return false;
+    }
+
+    if (bid.params.video.channel_id === undefined || bid.params.video.slot === undefined || bid.params.video.video_slot === undefined) {
+      utils.logMessage(BIDDER_CODE + ': channel_id, slot and/or video_slot are not present in bidder params');
+      return false;
+    }
+
+    const videoSlotDivId = utils.getBidIdParameter('video_slot', bid.params.video);
+    if (window.document.getElementById(videoSlotDivId) === null) {
+      utils.logMessage(BIDDER_CODE + ': video_slot HTML node id does not exist on the page');
+      return false;
+    }
+
+    const slotDivId = utils.getBidIdParameter('slot', bid.params.video);
+    if (window.document.getElementById(slotDivId) === null) {
+      utils.logMessage(BIDDER_CODE + ': slot HTML node id does not exist on the page');
+      return false;
+    }
+
+    return true;
+  },
+
+  /**
+   * Make a server request from the list of BidRequests.
+   * from Prebid.js: buildRequests - Takes an array of valid bid requests, all of which are guaranteed to have passed the isBidRequestValid() test.
+   *
+   * @param {BidRequest[]} bidRequests A non-empty list of bid requests which should be sent to the Server.
+   * @return ServerRequest Info describing the request to the server.
+   */
+  buildRequests: function(bidRequests, bidderRequest) {
+    const page = bidderRequest.refererInfo.referer;
+    const isPageSecure = !!page.match(/^https:/)
+
+    const siteId = '';
+    const bid = bidderRequest.bids[0];
+    const channelId = bid.params.video.channel_id;
+    let pubcid = null;
+
+    const videoSlotDiv = window.document.getElementById(utils.getBidIdParameter('video_slot', bid.params.video));
+    const hasVideoElement = videoSlotDiv.querySelectorAll('video').length;
+    const contentWidth = utils.getBidIdParameter('content_width', bid.params.video) ||
```

Comment:
This should be documented. The slot may be set to `display: none` and  thus the the `offsetWidth` is *0*. For DFP users this is often the case when slots are hidden before they are displayed.